conn_js.go

```diff
@@ -0,0 +1,196 @@
+package websocket
+
+import (
+	"bytes"
+	"errors"
+	"io"
+	"net"
+	"strings"
+	"sync"
+	"syscall/js"
+	"time"
+)
+
+const (
+	webSocketStateConnecting = 0
+	webSocketStateOpen       = 1
+	webSocketStateClosing    = 2
+	webSocketStateClosed     = 3
+)
+
+var errIsClosed = errors.New("connection is closed")
+
+// Conn implements net.Conn interface for WebSockets in js/wasm.
+type Conn struct {
+	js.Value
+	messageHandler *js.Func
+	closeHandler   *js.Func
+	mut            sync.Mutex
+	incomingData   chan []byte
+	currData       *bytes.Buffer
+	closeSignal    chan struct{}
+}
+
+func (c *Conn) Read(b []byte) (int, error) {
+	c.mut.Lock()
+	defer c.mut.Unlock()
+	if err := c.checkOpen(); err != nil {
+		return 0, io.EOF
+	}
+
+	if c.currData == nil {
+		select {
+		case data := <-c.incomingData:
+			c.currData = bytes.NewBuffer(data)
+		case <-c.closeSignal:
+			return 0, io.EOF
+		}
+	}
+
+	n, err := c.currData.Read(b)
+	if err == io.EOF {
+		c.currData = nil
+		return n, nil
+	}
+	return n, err
+}
+
+func (c *Conn) checkOpen() error {
+	state := c.Get("readyState").Int()
+	switch state {
+	case webSocketStateClosed, webSocketStateClosing:
+		return errIsClosed
+	}
+	return nil
+}
+
+func (c *Conn) Write(b []byte) (n int, err error) {
+	// c.mut.Lock()
+	// defer c.mut.Unlock()
+	if err := c.checkOpen(); err != nil {
+		return 0, err
+	}
+	c.Call("send", string(b))
+	return len(b), nil
+}
+
+// Close closes the connection. Only the first call to Close will receive the
+// close error, subsequent and concurrent calls will return nil.
+// This method is thread-safe.
+func (c *Conn) Close() error {
+	// c.mut.Lock()
+	// defer c.mut.Unlock()
+	c.Call("close")
+	if c.messageHandler != nil {
+		c.Call("removeEventListener", "message", *c.messageHandler)
+		c.messageHandler.Release()
+	}
+	if c.closeHandler != nil {
+		c.Call("removeEventListener", "close", *c.closeHandler)
+		c.closeHandler.Release()
+	}
+	return nil
+}
+
+func (c *Conn) LocalAddr() net.Addr {
+	// TODO(albrow): is there some way to get the local address?
+	return NewAddr("0.0.0.0:0")
```

Comment:
AFAIK the JavaScript WebSocket API does not expose any way to get the local address/port. Is that going to be a problem? I'm using the workaround of just returning `0.0.0.0:0` but maybe there is a better way to indicate that the local address is unknown?

Reply:
This should be fine.